## Упражнения по библиотеке Numpy

In [69]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [70]:
random_array = np.random.randint(-10, 10, size=100)
random_array

array([ -9, -10,   1,   1,   6,  -1,   5,   4,   4,   8,   1,   9,  -8,
        -6,   8,  -4,  -2,  -4,   7,  -7,   3,   7,  -2,  -9,   9,   4,
        -4,   1,  -3,   4,  -8,   3,   6,  -7,   7,  -3,  -7,  -9,  -5,
        -1,  -7,   7,   1,  -9,  -1,  -7,   3,   5,   4,  -3,   3,  -3,
         5,   2,   7,   4,   2,  -2,   4,   2, -10,  -4,  -2, -10,   1,
        -3,   0,   8,   6,  -3,  -8,  -8, -10,  -6,  -1,  -4,  -2,  -4,
        -2,  -3,   1,  -9, -10,   5,  -6,  -8,   1,  -3,  -8, -10,  -8,
        -6,   4,   3,  -8, -10,  -6,   3,  -4,  -2], dtype=int32)

In [71]:
random_array[(random_array > 3) & (random_array < 8)] *= -1
random_array

array([ -9, -10,   1,   1,  -6,  -1,  -5,  -4,  -4,   8,   1,   9,  -8,
        -6,   8,  -4,  -2,  -4,  -7,  -7,   3,  -7,  -2,  -9,   9,  -4,
        -4,   1,  -3,  -4,  -8,   3,  -6,  -7,  -7,  -3,  -7,  -9,  -5,
        -1,  -7,  -7,   1,  -9,  -1,  -7,   3,  -5,  -4,  -3,   3,  -3,
        -5,   2,  -7,  -4,   2,  -2,  -4,   2, -10,  -4,  -2, -10,   1,
        -3,   0,   8,  -6,  -3,  -8,  -8, -10,  -6,  -1,  -4,  -2,  -4,
        -2,  -3,   1,  -9, -10,  -5,  -6,  -8,   1,  -3,  -8, -10,  -8,
        -6,  -4,   3,  -8, -10,  -6,   3,  -4,  -2], dtype=int32)

**2.** Заменить максимальный элемент случайного массива на 0

In [72]:
max_el = random_array.max()
random_array[random_array == max_el]=0
random_array

array([ -9, -10,   1,   1,  -6,  -1,  -5,  -4,  -4,   8,   1,   0,  -8,
        -6,   8,  -4,  -2,  -4,  -7,  -7,   3,  -7,  -2,  -9,   0,  -4,
        -4,   1,  -3,  -4,  -8,   3,  -6,  -7,  -7,  -3,  -7,  -9,  -5,
        -1,  -7,  -7,   1,  -9,  -1,  -7,   3,  -5,  -4,  -3,   3,  -3,
        -5,   2,  -7,  -4,   2,  -2,  -4,   2, -10,  -4,  -2, -10,   1,
        -3,   0,   8,  -6,  -3,  -8,  -8, -10,  -6,  -1,  -4,  -2,  -4,
        -2,  -3,   1,  -9, -10,  -5,  -6,  -8,   1,  -3,  -8, -10,  -8,
        -6,  -4,   3,  -8, -10,  -6,   3,  -4,  -2], dtype=int32)

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [73]:
arrays = np.array([[1, 2], [3, 4], [5, 6]])
meshes = np.meshgrid(*arrays, indexing='ij')
result = np.stack(meshes, axis=-1).reshape(-1, arrays.shape[0])
result

array([[1, 3, 5],
       [1, 3, 6],
       [1, 4, 5],
       [1, 4, 6],
       [2, 3, 5],
       [2, 3, 6],
       [2, 4, 5],
       [2, 4, 6]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [75]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [1, 4, 7],
    [2, 5, 8],
    [3, 6, 9],
    [1, 5, 9],
    [3, 5, 7]
])

B = np.array([
    [1, 5],
    [3, 7]
])
result = []

for row in B:
    mask = np.array([np.all(np.isin(row, a)) for a in A])
    result.append(mask)

res_mask = np.any(result, axis=0)

matching_rows = A[res_mask]
matching_rows

array([[1, 5, 9],
       [3, 5, 7]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [77]:
matrix = np.random.randint(1, 3, size=(10, 3))
matrix

array([[2, 2, 2],
       [1, 2, 1],
       [1, 2, 2],
       [2, 2, 2],
       [2, 2, 2],
       [1, 2, 2],
       [1, 2, 1],
       [1, 2, 1],
       [1, 1, 1],
       [2, 1, 2]], dtype=int32)

In [78]:
def unique_rows(mtrx):
    unique_rows = mtrx[~np.all(mtrx == mtrx[:, [0]], axis=1)]
    return unique_rows
res = unique_rows(matrix)
res

array([[1, 2, 1],
       [1, 2, 2],
       [1, 2, 2],
       [1, 2, 1],
       [1, 2, 1],
       [2, 1, 2]], dtype=int32)

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [79]:
arr = np.random.randint(1,3, size=(15,3))
arr

array([[1, 1, 1],
       [1, 2, 2],
       [1, 1, 2],
       [1, 1, 1],
       [2, 2, 2],
       [1, 1, 2],
       [2, 2, 2],
       [1, 2, 1],
       [2, 1, 2],
       [2, 2, 2],
       [1, 2, 1],
       [1, 1, 1],
       [2, 1, 1],
       [1, 2, 2],
       [2, 2, 1]], dtype=int32)

In [80]:
unique_arr = np.unique(arr, axis=0)
unique_arr

array([[1, 1, 1],
       [1, 1, 2],
       [1, 2, 1],
       [1, 2, 2],
       [2, 1, 1],
       [2, 1, 2],
       [2, 2, 1],
       [2, 2, 2]], dtype=int32)

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [81]:
X = [[1, 0, 1], 
     [2, 0, 2], 
     [3, 0, 3], 
     [4, 4, 4]]
res = 1
for i in range(min(len(X), len(X[0]))):  
        if X[i][i] != 0:  
            res *= X[i][i]  
res

3

In [82]:
 X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
diagonal_elements = np.diagonal(X)
nonzero_elements = diagonal_elements[diagonal_elements != 0]
res = np.prod(nonzero_elements) 
int(res)

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [83]:
x = [1,2,2,4]
y = [4,2,1,2]
def is_same_multiset(x,y):
    return sorted(x)==sorted(y)
is_same_multiset(x,y)

True

In [84]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 5])
def is_same_multiset(x, y):
    # Получаем уникальные элементы и их количество для каждого вектора
    unique_x, counts_x = np.unique(x, return_counts=True)
    unique_y, counts_y = np.unique(y, return_counts=True)
    
    # Сравниваем уникальные элементы и их количество
    return np.array_equal(unique_x, unique_y) and np.array_equal(counts_x, counts_y)
is_same_multiset(x,y)

False

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [85]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max_el = 0
for i in range (len(x)):
    if x[i] == 0 and i < len(x)-1:
        if max_el < x[i+1]:
            max_el = x[i+1]
max_el

5

In [86]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
def max_after_zero(x):
    zero_indices = np.where(x == 0)[0]
    elements_after_zero = []
    for idx in zero_indices:
        if idx < len(x)-1:  
            elements_after_zero.append(x[idx + 1])
    return np.max(elements_after_zero) if elements_after_zero else None
res = int(max_after_zero(x))
res

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [87]:
x = [2,2,2,3,3,3,5]
def run_length_encoding(x):
    values = []
    counts = []
    prev = x[0]
    count = 1
    
    for i in range(1, len(x)):
        if x[i] == prev:
            count += 1
        else:
            values.append(prev)
            counts.append(count)
            prev = x[i]
            count = 1
    values.append(prev)
    counts.append(count)
    return values, counts

values, counts = run_length_encoding(x)
values, counts

([2, 3, 5], [3, 3, 1])

In [88]:
def run_length_encoding(x):
    values, counts = np.unique(x, return_counts=True)
    return values, counts
values, counts = run_length_encoding(x)
values, counts

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [89]:
import time
from scipy.spatial.distance import cdist
np.random.seed(42)
X = np.random.randint(0, 100, (5, 2))
Y = np.random.randint(0, 100, (5, 2))

start = time.time()
m, n = len(X), len(Y)
dist_matrix = [[sum((X[i][k] - Y[j][k]) ** 2 for k in range(len(X[0]))) ** 0.5 for j in range(n)] for i in range(m)]
result = np.array(dist_matrix)
manual_time = time.time() - start

start = time.time()
scipy_result = cdist(X, Y, metric='euclidean')
scipy_time = time.time() - start

In [90]:
X

array([[51, 92],
       [14, 71],
       [60, 20],
       [82, 86],
       [74, 74]], dtype=int32)

In [91]:
Y

array([[87, 99],
       [23,  2],
       [21, 52],
       [ 1, 87],
       [29, 37]], dtype=int32)

In [92]:
result

array([[ 36.67424164,  94.25497334,  50.        ,  50.24937811,
         59.23681288],
       [ 78.18567644,  69.58448103,  20.24845673,  20.61552813,
         37.16180835],
       [ 83.48652586,  41.14608122,  50.44799302,  89.27485648,
         35.35533906],
       [ 13.92838828, 102.6498904 ,  69.83552105,  81.0061726 ,
         72.18032973],
       [ 28.17800561,  88.23264702,  57.38466694,  74.14849965,
         58.25804665]])

In [93]:
scipy_result

array([[ 36.67424164,  94.25497334,  50.        ,  50.24937811,
         59.23681288],
       [ 78.18567644,  69.58448103,  20.24845673,  20.61552813,
         37.16180835],
       [ 83.48652586,  41.14608122,  50.44799302,  89.27485648,
         35.35533906],
       [ 13.92838828, 102.6498904 ,  69.83552105,  81.0061726 ,
         72.18032973],
       [ 28.17800561,  88.23264702,  57.38466694,  74.14849965,
         58.25804665]])

In [94]:
manual_time

0.0007140636444091797

In [95]:
scipy_time

0.00024700164794921875

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [96]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [97]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [98]:
average_calories = float(np.mean(calorie_stats))
average_calories

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [99]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [100]:
median_calories = float(np.median(calorie_stats_sorted))
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [101]:
nth_percentile = float(np.percentile(calorie_stats_sorted, 4))
nth_percentile

70.0

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [102]:
more_calories = float((np.sum(calorie_stats_sorted > 60)/calorie_stats_sorted.size)*100)
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [103]:
calorie_std = float(np.std(calorie_stats_sorted))
calorie_std

19.35718533390827

Разброс небольшой, а это значит, что можно сделать обобщение, что в большинстве злаков содержится около 100 калорий

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Наглядно видно, что практически во всех хлопьях (96%) больше калорий, чем в CrunchieMunchies, поэтому можно сделать акцент на этом в маркетинге, продвигать их на аудиторию, которая следит за питанием, так как это хорошая характеристика, которая выгодно отличает эти хлопья от других и определенно привлечет покупателей.